### PR TITLE
registry: Allow the caller to pass the sample.

### PR DESCRIPTION
**Pull request by mccomplete**

### CODE REVIEW

**Review by dgilboa — CHANGES_REQUESTED** [review]
Can you explain why it is needed for ? 
Any UT ?

File: pyformance/registry.py [hunk 1]
```diff
@@ -126,7 +126,7 @@ def meter(self, key, tags=None):
     def create_sink(self):
         return None
 
-    def timer(self, key, tags=None):
+    def timer(self, key, tags=None, sample=None):
```
Comment (dgilboa) [inline pyformance/registry.py:129]:
```suggestion
    def timer(self, key, tags=None, sample_manager=None):
```

File: pyformance/registry.py [hunk 2]
```diff
@@ -136,6 +136,9 @@ def timer(self, key, tags=None):
         :param tags: tags attached to the timer (e.g. {'region': 'us-west-1'})
         :type tags: C{dict}
 
+        :param sample: a sample manager, default is ExpDecayingSample
```
Comment (dgilboa) [inline pyformance/registry.py:139]: consider stating this default in the function definition itself and not only in the comment.

**Comment by dgilboa** [comment]
@mccomplete please state in the commit message that this actually allows to use a different sampling strategy (or alike) for a timer metric (did I understand correctly ?)

**Comment by dgilboa** [comment]
@ilya-lt would be great if u can also have a look here as you know this lib better than I do.

File: pyformance/registry.py (hunk 1 above)
Comment (mccomplete) [reply pyformance/registry.py:129]: I was going for consistency with the library as it is. The property of `Timer` and the argument to the `Timer` constructor (and `Histogram` etc.) are called `sample`. If I change it here, should I change it all the way down?

**Comment by dgilboa** [comment]
@mccomplete ping when you're done.

File: pyformance/registry.py (hunk 1 above)
Comment (dgilboa) [reply pyformance/registry.py:129]: sounds like it is the correct way to go but if that's the case than keeping it in the same convention is ok for us now I suppose. 
resolving.

**Comment by mccomplete** [comment]
@dgilboa 

> Can you explain why it is needed for ?

The influx reporter periodically pushes metrics to influx with the current timestamp. All the metrics managed by the sample manager are sent with the current timestamp, no matter when the metrics may have been actually measured. 

With the decaying sample, we often see metrics that are very old in Grafana, which makes no sense for our use case. 

The sliding time window sample gives us more or less what we want. 

We might be able to use events instead of samples (actually, instead of timers), but that would require us to rewrite the way we use metrics, and probably do a lot of calculations on the influx side. It wouldn't be pretty.

**Comment by dgilboa** [comment]
@mccomplete - as requested - just amend the commit msg and add the explanation - I think this should be enough as I wrote earlier.

File: pyformance/registry.py (hunk 2 above)
Comment (mccomplete) [reply pyformance/registry.py:139]: These classes are mutable, so it wouldn't be a good idea to use them as defaults in the function definition

https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

**Comment by mccomplete** [comment]
@dgilboa 

> please state in the commit message that this actually allows to use a different sampling strategy (or alike) for a timer metric (did I understand correctly ?)

Done, PTAL

**Review by dgilboa — APPROVED** [review]

**Pull request merged**